src/core/qgsvectorlayer.cpp

```diff
@@ -5323,6 +5323,17 @@ void QgsVectorLayer::emitDataChanged()
   mDataChangedFired = false;
 }
 
+void QgsVectorLayer::onAfterCommitChangesDependency()
+{
+  // Refreshing the data provider right away causes problems on a layer dependent on itself.
+  // In that case two `dataChanged` signals are expected to be emitted when deleting or creating
+  // a feature, but as the layer is reloaded, the second `dataChanged` is not emitted.
```

Comment:
would it work to test if `mDataChangedFired` is true ?

Reply:
Nice idea, it works! Thanks!